src/systems/filecoin_markets/storage_market/storage_market_actor_code.go

```diff
@@ -7,82 +7,60 @@ import (
 	actor "github.com/filecoin-project/specs/systems/filecoin_vm/actor"
 	addr "github.com/filecoin-project/specs/systems/filecoin_vm/actor/address"
 	ai "github.com/filecoin-project/specs/systems/filecoin_vm/actor_interfaces"
-	vmr "github.com/filecoin-project/specs/systems/filecoin_vm/runtime"
 	util "github.com/filecoin-project/specs/util"
 )
 
 ////////////////////////////////////////////////////////////////////////////////
 // Actor methods
 ////////////////////////////////////////////////////////////////////////////////
 
-func (a *StorageMarketActorCode_I) WithdrawBalance(rt Runtime, minerAddr addr.Address, amountRequested actor.TokenAmount) {
+func (a *StorageMarketActorCode_I) WithdrawBalance(rt Runtime, entryAddr addr.Address, amountRequested actor.TokenAmount) {
 	IMPL_FINISH() // BigInt arithmetic
 	amountSlashedTotal := actor.TokenAmount(0)
 
-	TODO() // TODO: must also handle case of non-miner (client) sender
-
-	ownerAddr, workerAddr := _rtGetMinerAccountsAssert(rt, minerAddr)
-	rt.ValidateImmediateCallerInSet([]addr.Address{ownerAddr, workerAddr})
-	TODO() // TODO: should workerAddr be permitted here? Depends on actor principal security assumptions.
-
-	callerAddr := rt.ImmediateCaller()
-
 	if amountRequested < 0 {
-		rt.AbortArgMsg("sma.WithdrawBalance: negative amount.")
+		rt.AbortArgMsg("Negative amount.")
 	}
 
-	h, st := a.State(rt)
+	recipientAddr := _rtValidateImmediateCallerDetermineRecipient(rt, entryAddr)
 
-	if !st._addressEntryExists(minerAddr) {
-		rt.AbortArgMsg("sma.WithdrawBalance: address entry does not exist")
-	}
+	h, st := a.State(rt)
+	st._rtAbortIfAddressEntryDoesNotExist(rt, entryAddr)
 
 	// Before any operations that check the balance tables for funds, execute all deferred
 	// deal state updates.
 	//
 	// Note: as an optimization, implementations may cache efficient data structures indicating
 	// which of the following set of updates are redundant and can be skipped.
-	amountSlashedTotal += st._rtUpdatePendingDealStatesForParty(rt, minerAddr)
+	amountSlashedTotal += st._rtUpdatePendingDealStatesForParty(rt, entryAddr)
 
-	minBalance := st._getLockedReqBalanceInternal(minerAddr)
+	minBalance := st._getLockedReqBalanceInternal(entryAddr)
```

Comment:
Lets confirm the math here is correct. It needs to first deduct amountSlashedTotal from the BalanceTable and then extract amount no?

Reply:
No, I don't think so -- `_rtUpdatePendingDealStatesForParty` mutates the balance table in-place, so it should already reflect the slashed amounts.